File: 백준/Gold/15486. 퇴사 2/퇴사 2.py
import sys

def input():
    return sys.stdin.readline().rstrip()

cost_lst = [(-1,-1)]
N = int(input())
dp = [0 for _ in range(N+1)]

for _ in range(N):
    T, P = map(int,input().split())
    cost_lst.append((T,P))
ans = 0
#i1은 일
for i0 in range(1,N+1):
    #이전에 가져온경우와 가지고오지 않은 경우중에 선택했는지 안했는지 둘다 고려한 코드라인
    dp[i0] = max(dp[i0],dp[i0-1])
    if i0 + cost_lst[i0][0]-1<= N:
        dp[i0 + cost_lst[i0][0]-1] = max(dp[i0 + cost_lst[i0][0] - 1],dp[i0-1]+cost_lst[i0][1])

print(dp[N])

File: 백준/Gold/15486. 퇴사 2/퇴사 2.java
import java.util.*;
import java.io.*;
public class Main {
    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int N = Integer.parseInt(br.readLine());
        int[] arr = new int[N+2];
        int[] val = new int[N+2];
        int[] dp = new int[N+2];
        int ans = -1;
        StringTokenizer st;
        for(int i=0;i<N;i++){
            st = new StringTokenizer(br.readLine());
            arr[i] = Integer.parseInt(st.nextToken());
            val[i] = Integer.parseInt(st.nextToken());
        }
        for(int i=N;i>=0;i--){
            if(i+arr[i]<=N){
                dp[i] = Math.max(val[i] + dp[i + arr[i]], dp[i+1]);
            }
            else{
                dp[i] = dp[i+1];
            }
        }
        System.out.println(dp[0]);
        }
}
